### PR TITLE
[Architecture] Establish capability-module ownership manifests

### PR DESCRIPTION
**Pull request by Z-M-Huang**
## Summary

Implements issue #358 by moving architecture ownership into independently owned capability fragments while preserving the existing production/evaluation/browser/worker discovery graph.

- Added `architecture/manifest.mjs` as the shared deterministic loader and module-boundary policy.
- Replaced the monolithic unit/exception/worker inventory with 51 `architecture/modules/<capability>.json` fragments plus the central inventory manifest.
- Added explicit public/private visibility, `postgres_adapter`, and `postgres_infrastructure` roles with the narrow adapter-to-infrastructure allowance.
- Remapped retained exception owners and worker lifecycle obligations to the approved follow-up issues; removed only the completed repository-to-PostgreSQL bridge exception.
- Added ADR 0005 refining ADRs 0003/0004 and expanded architecture conformance UAT coverage.

## Inventory reconciliation

- 57 Go units (production and evaluation)
- 40 reachable browser units
- 25 retained exceptions
- 52 worker anchors
- completed issues: 260, 347, 261, 262, 263, 348

## Audit and verification

- Independent `gpt-5.6-terra` / max plan audit against base `ee30a3c86305fdc325eec72569bea6673a646d24`: `conformant` (initial audit finding corrections were re-audited to a clean result).
- `node --test tests/uat/architecture_conformance.test.mjs`: 32/32 passed.
- `node scripts/check-architecture.mjs`: passed (57 Go packages, 40 browser modules).
- `./scripts/ci-check.sh`: passed, including 90.0% coverage.
- No runtime, persistence, migration, provider, semantic, ranking, or evaluation behavior changed; the approved deterministic evaluation waiver applies.
- Final-head production-image full-matrix E2E remains required through the repository automation gate.

## Risks and mitigations

- Fragment omissions or duplicate ownership could weaken enforcement: the loader rejects absent/empty, missing, unlisted, duplicate-reference, duplicate-capability, filename-mismatched, and misowned fragments, while discovery reconciles the merged graph.
- Infrastructure visibility could legalize policy bypasses: concrete adapters and PostgreSQL infrastructure are private; only PostgreSQL adapters may consume the explicitly classified infrastructure, with negative UAT coverage for application, transport, and other adapters.
- Compatibility obligations could outlive their owner: each retained exception/lifecycle anchor records one replacement issue and expires when that issue is completed.

## Scope

One commit, no generated evaluation/runtime/database artifacts staged. `coverage.out` was produced by local CI and is disposable under the workflow cleanup approval; pre-existing evaluation/runtime/browser artifacts are preserved.

## Summary by CodeRabbit

- **New Features**
  - Added capability-based architecture manifests covering application, domain, transport, adapter, portal, and composition areas.
  - Added automated discovery and validation for architecture ownership, visibility, dependency boundaries, exceptions, and worker lifecycle tracking.
  - Added clearer diagnostics for missing, duplicate, malformed, or incorrectly assigned architecture definitions.

- **Documentation**
  - Added an architecture decision record describing capability ownership, dependency rules, compatibility exceptions, and verification requirements.

- **Tests**
  - Expanded conformance coverage for manifest loading, inheritance, visibility enforcement, dependency reuse, and lifecycle obligations.

### CODE REVIEW

File: tests/uat/architecture_conformance.test.mjs [hunk 1]
```diff
@@ -56,6 +58,192 @@ function assertEdgeExpectation(edge, result) {
...
+    fs.copyFileSync(path.join(root, reference), path.join(fixtureRoot, reference));
+  }
+  return { fixtureRoot, rootManifest };
+}
+
+test("loads the complete independently owned architecture inventory", () => {
+  assert.equal(productionManifest.load_diagnostics.length, 0);
+  assert.equal(productionManifest.fragments.length, 51);
+  assert.equal(productionManifest.go.units.length, 57);
+  assert.equal(productionManifest.browser.units.length, 40);
```
Comment (MarkHuang-AI-Service) [inline tests/uat/architecture_conformance.test.mjs:76]: ✅ **Fixed in `9d5bf3867892b572cbf6d912b27dc076ac7ca962`**

A fresh AI inspection found that the original failure path is no longer present at this head.

---

🔴 High **Missing test for duplicate Go/browser unit across fragments**

**Why it matters:** The loader rejects duplicate Go or browser unit IDs across different fragments (manifest.mjs lines 356-362), which is critical for the independent ownership model. Without a test that places the same unit ID in two different fragments, a regression could allow duplicate ownership, undermining the entire fragment system. The existing 'duplicate capability' test only covers fragment-level capability duplicates, not unit-level duplicates across fragments.

**Fix:** Add a test that creates two fragments both declaring the same Go unit ID (e.g., 'github.com/markhuangai/dense-mem/internal/repository') and verifies the loader produces a 'duplicate-fragment' diagnostic mentioning the unit ID. Similarly test browser unit duplicates.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```text
Verify this finding against the current code. Fix it only if it is still valid,
keep the change minimal, and run the relevant tests.

Target: `@tests/uat/architecture_conformance.test.mjs:76`
Finding: Missing test for duplicate Go/browser unit across fragments
Impact: The loader rejects duplicate Go or browser unit IDs across different fragments (manifest.mjs lines 356-362), which is critical for the independent ownership model. Without a test that places the same unit ID in two different fragments, a regression could allow duplicate ownership, undermining the entire fragment system. The existing 'duplicate capability' test only covers fragment-level capability duplicates, not unit-level duplicates across fragments.
Requested fix: Add a test that creates two fragments both declaring the same Go unit ID (e.g., 'github.com/markhuangai/dense-mem/internal/repository') and verifies the loader produces a 'duplicate-fragment' diagnostic mentioning the unit ID. Similarly test browser unit duplicates.
```

</details>

Comment (MarkHuang-AI-Service) [inline tests/uat/architecture_conformance.test.mjs:76]: ✅ **Fixed in `9d5bf3867892b572cbf6d912b27dc076ac7ca962`**

A fresh AI inspection found that the original failure path is no longer present at this head.

---

🟠 Moderate **Missing test for fragment capability filename match**

**Why it matters:** The loader validates that fragment.capability matches its filename (manifest.mjs lines 340-342), preventing misconfiguration where a fragment named 'foo.json' declares capability: 'bar'. Without a test, a regression could allow mismatched capabilities, making the fragment system unpredictable and harder to reason about.

**Fix:** Add a test that creates a fragment file named 'foo.json' with capability: 'bar' and verifies the loader produces an 'invalid-fragment' diagnostic stating 'capability must match its filename foo'.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```text
Verify this finding against the current code. Fix it only if it is still valid,
keep the change minimal, and run the relevant tests.

Target: `@tests/uat/architecture_conformance.test.mjs:76`
Finding: Missing test for fragment capability filename match
Impact: The loader validates that fragment.capability matches its filename (manifest.mjs lines 340-342), preventing misconfiguration where a fragment named 'foo.json' declares capability: 'bar'. Without a test, a regression could allow mismatched capabilities, making the fragment system unpredictable and harder to reason about.
Requested fix: Add a test that creates a fragment file named 'foo.json' with capability: 'bar' and verifies the loader produces an 'invalid-fragment' diagnostic stating 'capability must match its filename foo'.
```

</details>

File: tests/uat/architecture_conformance.test.mjs [hunk 2]
```diff
@@ -56,6 +58,192 @@ function assertEdgeExpectation(edge, result) {
...
+    ));
+    duplicateFragment.capability = "architecture";
+    fs.writeFileSync(
+      path.join(duplicateCapability.fixtureRoot, "architecture/modules/architecture-copy.json"),
+      JSON.stringify(duplicateFragment, null, 2),
+    );
+    duplicateCapability.rootManifest.fragments.push("architecture/modules/architecture-copy.json");
+    fs.writeFileSync(
+      path.join(duplicateCapability.fixtureRoot, "architecture/ownership.v1.json"),
+      JSON.stringify(duplicateCapability.rootManifest, null, 2),
```
Comment (MarkHuang-AI-Service) [inline tests/uat/architecture_conformance.test.mjs:185]: ✅ **Fixed in `9d5bf3867892b572cbf6d912b27dc076ac7ca962`**

A fresh AI inspection found that the original failure path is no longer present at this head.

---

🟠 Moderate **Missing test for unit capability inheritance enforcement**

**Why it matters:** The loader explicitly rejects Go and browser units that declare a capability field in fragments (manifest.mjs lines 374-376 and 380-382), enforcing that units must inherit capability from their fragment. Without a negative test, a regression could allow units to declare capabilities, violating the inheritance invariant and making capability ownership ambiguous.

**Fix:** Add a test that creates a fragment with a Go unit declaring capability: 'foo' and verifies the loader produces an 'invalid-fragment' diagnostic stating the unit 'must inherit capability'. Repeat for browser units.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```text
Verify this finding against the current code. Fix it only if it is still valid,
keep the change minimal, and run the relevant tests.

Target: `@tests/uat/architecture_conformance.test.mjs:170`
Finding: Missing test for unit capability inheritance enforcement
Impact: The loader explicitly rejects Go and browser units that declare a capability field in fragments (manifest.mjs lines 374-376 and 380-382), enforcing that units must inherit capability from their fragment. Without a negative test, a regression could allow units to declare capabilities, violating the inheritance invariant and making capability ownership ambiguous.
Requested fix: Add a test that creates a fragment with a Go unit declaring capability: 'foo' and verifies the loader produces an 'invalid-fragment' diagnostic stating the unit 'must inherit capability'. Repeat for browser units.
```

</details>

File: tests/uat/architecture_conformance.test.mjs [hunk 3]
```diff
@@ -56,6 +58,192 @@ function assertEdgeExpectation(edge, result) {
...
+    fs.rmSync(duplicateCapability.fixtureRoot, { recursive: true, force: true });
+  }
+
+  const misplaced = copyManifestFixture();
+  try {
+    const sourceFragmentPath = path.join(misplaced.fixtureRoot, "architecture/modules/http-transport.json");
+    const sourceFragment = JSON.parse(fs.readFileSync(sourceFragmentPath, "utf8"));
+    const architecturePath = path.join(misplaced.fixtureRoot, "architecture/modules/architecture.json");
+    const architectureFragment = JSON.parse(fs.readFileSync(architecturePath, "utf8"));
+    architectureFragment.exceptions.push(sourceFragment.exceptions[0]);
```
Comment (MarkHuang-AI-Service) [inline tests/uat/architecture_conformance.test.mjs:199]: ✅ **Fixed in `9d5bf3867892b572cbf6d912b27dc076ac7ca962`**

A fresh AI inspection found that the original failure path is no longer present at this head.

---

🟠 Moderate **Missing test for worker-to-worker and same-capability private imports**

**Why it matters:** The evaluateModuleEdge function has two special cases: worker-to-worker imports are always allowed (manifest.mjs line 247), and same-capability private imports are allowed (line 254). The existing visibility test covers cross-capability private/public scenarios but not these two critical cases. Without tests, incorrect implementations of these special cases could either break valid imports or allow forbidden ones.

**Fix:** Add test cases to 'enforces private visibility and narrow PostgreSQL infrastructure reuse': (1) worker-to-worker private import across capabilities should be allowed, (2) same-capability private import should be allowed even when both units are private.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```text
Verify this finding against the current code. Fix it only if it is still valid,
keep the change minimal, and run the relevant tests.

Target: `@tests/uat/architecture_conformance.test.mjs:184`
Finding: Missing test for worker-to-worker and same-capability private imports
Impact: The evaluateModuleEdge function has two special cases: worker-to-worker imports are always allowed (manifest.mjs line 247), and same-capability private imports are allowed (line 254). The existing visibility test covers cross-capability private/public scenarios but not these two critical cases. Without tests, incorrect implementations of these special cases could either break valid imports or allow forbidden ones.
Requested fix: Add test cases to 'enforces private visibility and narrow PostgreSQL infrastructure reuse': (1) worker-to-worker private import across capabilities should be allowed, (2) same-capability private import should be allowed even when both units are private.
```

</details>

**Review by MarkHuang-AI-Service — COMMENTED** [review]
## 🔎 AI review

Found **4 actionable issues** — 1 high and 3 moderate.

See the inline comments for details.

<details>
<summary>📊 Token usage · 4,055,101 tokens</summary>

| Model | Input | Output | Cache hit | Cache creation | Total |
|:--|--:|--:|--:|--:|--:|
| <code>auto</code> | 1,666,802 | 53,141 | 2,335,158 | 0 | **4,055,101** |

> ✅ **Complete SDK accounting**
> ℹ️ Claude Agent SDK model usage only; external MCP service usage is excluded.

</details>

**Review by MarkHuang-AI-Service — COMMENTED** [review]
## ⚠️ Stale AI review

This review covered `9d5bf3867892b572cbf6d912b27dc076ac7ca962`. New changes were pushed in `10a2193498c0735620a4c39bc998ac114848517b`, so this result is no longer current.

**Review by MarkHuang-AI-Service — COMMENTED** [review]
## ⚠️ Stale AI review

This review covered `10a2193498c0735620a4c39bc998ac114848517b`. New changes were pushed in `afb56845e2cfa1b6a088989a98634398ed865e30`, so this result is no longer current.

**Review by MarkHuang-AI-Service — COMMENTED** [review]
## ⚠️ Stale AI review

This review covered `afb56845e2cfa1b6a088989a98634398ed865e30`. New changes were pushed in `5a9df5dbfcccdeef47448e3a889c9dd3ba951579`, so this result is no longer current.

**Review by MarkHuang-AI-Service — COMMENTED** [review]
## ⚠️ Stale AI review

This review covered `5a9df5dbfcccdeef47448e3a889c9dd3ba951579`. New changes were pushed in `6a07fcb301ef46431fb93d9877d06f523dd76c56`, so this result is no longer current.

File: architecture/manifest.mjs [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+    diagnostics.push(diagnostic("invalid-fragment", `${relativePath} must contain an object`));
+    return;
+  }
+  if (fragment.schema_version !== 1) {
+    diagnostics.push(diagnostic("invalid-fragment", `${relativePath} schema_version must be 1`));
+  }
+  if (typeof fragment.capability !== "string" || fragment.capability.length === 0 || hasWildcard(fragment.capability)) {
+    diagnostics.push(diagnostic("invalid-fragment", `${relativePath} needs a concrete capability`));
+  }
+  for (const key of ["completed_issues", "exceptions", "workers"]) {
```
[the remark this answers is not in the document]
Comment (Z-M-Huang) [reply architecture/manifest.mjs:368]: Fixed in 5a9df5db: the fragment loader now rejects central-only fields (`module`, `allowed_targets`, `fragments`, `go.profiles`, `browser.entries`, and `browser.exclusions`) instead of silently discarding them. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: architecture/manifest.mjs [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+        diagnostics.push(diagnostic("invalid-fragment", `${record.ref} worker ${worker.path} must be owned by ${owner.ref}`));
+      }
+    }
+  }
+
+  const merged = {
+    schema_version: rootManifest.schema_version,
+    module: rootManifest.module,
+    allowed_targets: rootManifest.allowed_targets,
+    completed_issues: completedIssues,
```
[the remark this answers is not in the document]
Comment (Z-M-Huang) [reply architecture/manifest.mjs:577]: Fixed in 9d5bf386: loadManifest now rejects root fragment-owned completion records before merging fragments, and the architecture UAT covers root `completed_issues` and `enforced_through_issue`. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: architecture/manifest.mjs [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+        diagnostics.push(diagnostic("wildcard", `${kind} unit ${unit.id} must be exact`));
+      }
+      if (!roles.has(unit.role)) {
+        diagnostics.push(diagnostic("unknown-role", `${kind} unit ${unit.id} uses ${unit.role ?? "no role"}`));
+      }
+      if (typeof unit.capability !== "string" || unit.capability.length === 0 || hasWildcard(unit.capability)) {
+        diagnostics.push(diagnostic("invalid-manifest", `${kind} unit ${unit.id} needs a concrete capability`));
+      }
+      if (!isVisibility(unit.visibility)) {
+        diagnostics.push(diagnostic("invalid-manifest", `${kind} unit ${unit.id} needs visibility public or private`));
```
[the remark this answers is not in the document]
Comment (Z-M-Huang) [reply architecture/manifest.mjs:165]: Fixed in 9d5bf386: validateManifest enforces private visibility for `postgres_adapter` and `postgres_infrastructure` units, with a negative public-infrastructure test. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: architecture/manifest.mjs [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+  return diagnostics;
+}
+
+export function evaluateModuleEdge(sourceUnit, targetUnit) {
+  if (!sourceUnit || !targetUnit) {
+    return { ok: false, diagnostic: diagnostic("unclassified", "module edge names an unclassified unit") };
+  }
+  const roleAllowed = sourceUnit.role === targetUnit.role && sourceUnit.role === "worker"
+    ? true
+    : allowedTargets[sourceUnit.role]?.includes(targetUnit.role);
```
[the remark this answers is not in the document]
Comment (Z-M-Huang) [reply architecture/manifest.mjs]: Fixed in 10a21934: evaluateModuleEdge now applies the declared `allowedTargets.worker` matrix without a worker-to-worker bypass, and the UAT proves cross-capability worker imports are forbidden. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: architecture/manifest.mjs [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+      exceptions: [],
+      workers: [],
+      load_diagnostics: [diagnostic("invalid-manifest", "root manifest must be an object")],
+    };
+  }
+
+  const rootFragmentOwnedFields = [];
+  for (const key of ["completed_issues", "exceptions", "workers"]) {
+    if (Object.prototype.hasOwnProperty.call(rootManifest, key)) {
+      rootFragmentOwnedFields.push(diagnostic("invalid-fragment", `root manifest must not define ${key}; move it to its capability fragment`));
```
[the remark this answers is not in the document]
Comment (Z-M-Huang) [reply architecture/manifest.mjs:413]: Fixed in afb56845: loadManifest rejects obsolete `enforced_through_issue` in both the root manifest and capability fragments, with UAT coverage for each location. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: tests/uat/architecture_conformance.test.mjs (hunk 1 above)
Comment (Z-M-Huang) [reply tests/uat/architecture_conformance.test.mjs:76]: Fixed in 9d5bf386: the loader rejects duplicate Go/browser unit ownership across fragments, with the architecture UAT covering both duplicate unit cases. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

Comment (Z-M-Huang) [reply tests/uat/architecture_conformance.test.mjs:76]: Fixed in 9d5bf386: the loader validates that each capability fragment name matches its declared capability, with a negative filename/capability mismatch test. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: tests/uat/architecture_conformance.test.mjs (hunk 2 above)
Comment (Z-M-Huang) [reply tests/uat/architecture_conformance.test.mjs:185]: Fixed in 9d5bf386: fragment Go and browser units now inherit the fragment capability and cannot override it, with negative inheritance tests for both unit kinds. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

File: tests/uat/architecture_conformance.test.mjs (hunk 3 above)
Comment (Z-M-Huang) [reply tests/uat/architecture_conformance.test.mjs:199]: Fixed in 9d5bf386 and 10a21934: the UAT covers same-capability private imports and rejects cross-capability worker-to-worker imports according to the role matrix. `node --test tests/uat/architecture_conformance.test.mjs` passes 37/37.

**Review by MarkHuang-AI-Service — COMMENTED** [review]
## ✨ Good job!

No actionable issues found.

<details>
<summary>📊 Token usage · 4,564,521 tokens</summary>

| Model | Input | Output | Cache hit | Cache creation | Total |
|:--|--:|--:|--:|--:|--:|
| <code>auto</code> | 1,541,595 | 87,854 | 2,935,072 | 0 | **4,564,521** |

> ✅ **Complete SDK accounting**
> ℹ️ Claude Agent SDK model usage only; external MCP service usage is excluded.

</details>
